packages/preview-middleware/src/base/config.ts

```diff
@@ -0,0 +1,396 @@
+    return {
+        basePath: posix.relative(posix.dirname(config.path), '/') ?? '.',
+        apps: {},
+        init: config.init ? ns + config.init : undefined,
+        ui5: {
+            libs: getUI5Libs(manifest),
+            theme: ui5Theme,
+            flex,
+            resources: {
+                [PREVIEW_URL.client.ns]: PREVIEW_URL.client.url
+            },
+            bootstrapOptions: ''
+        },
+        locateReuseLibsScript: config.libs
+    };
+}
+
+/**
+ * Creates the configuration object for the test template.
+ *
+ * @param config test configuration
+ * @param id application id
+ * @returns configuration object for the test template
+ */
+export function createTestTemplateConfig(config: InternalTestConfig, id: string) {
+    return {
+        id,
+        framework: config.framework,
+        basePath: posix.relative(posix.dirname(config.path), '/') ?? '.',
+        initPath: posix.relative(posix.dirname(config.path), config.init)
+    };
+}
+
+export type PreviewUrls = {
+    path: string;
+    type: 'preview' | 'editor' | 'test';
+};
+
+/**
+ * Returns the preview paths.
+ *
+ * @param config configuration from the ui5.yaml
+ * @param logger logger instance
+ * @returns an array of preview paths
+ */
+export function getPreviewPaths(config: MiddlewareConfig, logger: ToolsLogger = new ToolsLogger()): PreviewUrls[] {
+    const urls: PreviewUrls[] = [];
+    // remove incorrect configurations
+    sanitizeConfig(config, logger);
+    // add flp preview url
+    const flpConfig = getFlpConfigWithDefaults(config.flp);
+    urls.push({ path: `${flpConfig.path}#${flpConfig.intent.object}-${flpConfig.intent.action}`, type: 'preview' });
+    // add editor urls
+    if (config.rta?.editors) {
+        config.rta.editors.forEach((editor) => {
+            urls.push({ path: editor.path, type: 'editor' });
+        });
+    }
+    // add test urls if configured
+    if (config.test) {
+        config.test.forEach((test) => {
+            const testConfig = mergeTestConfigDefaults(test);
+            urls.push({ path: testConfig.path, type: 'test' });
+        });
+    }
+    return urls;
+}
+
+/**
+ * Generates the preview files.
+ *
+ * @param basePath path to the application root
+ * @param config configuration from the ui5.yaml
+ * @param fs file system editor
+ * @param logger logger instance
+ * @returns a mem-fs editor with the preview files
+ */
+export async function generatePreviewFiles(
+    basePath: string,
+    config: MiddlewareConfig,
+    fs?: Editor,
+    logger: ToolsLogger = new ToolsLogger()
+) {
+    // remove incorrect configurations
+    sanitizeConfig(config, logger);
+
+    // create file system if not provided
+    if (!fs) {
+        fs = create(createStorage());
+    }
+
+    const templatePath = join(__dirname, '../../templates');
+    const webappPath = await getWebappPath(basePath, fs);
+    const manifest = (await fs.readJSON(join(webappPath, 'manifest.json'))) as unknown as Manifest;
```

Comment:
I do not understand the following: how does the manifest content get into the fs storage? As fs is just created a few steps before.

Reply:
`mem-fs` and `mem-fs-editor` work as an overlay over the filesystem.